modules/core/src/main/scala/org/tessellation/infrastructure/snapshot/GlobalSnapshotStorage.scala

```diff
@@ -60,6 +65,40 @@ object GlobalSnapshotStorage {
       }
   }
 
+  def make[F[_]: KryoSerializer](
+    ref: MapRef[F, SnapshotOrdinal, Option[GlobalSnapshot]]
+  )(implicit F: Async[F]): GlobalSnapshotStorage[F] =
+    new GlobalSnapshotStorage[F] {
+
+      implicit val ordinalOrder: Order[SnapshotOrdinal] = Order[Long].contramap[SnapshotOrdinal](_.value.value)
+      implicit val ordinalOrdering = ordinalOrder.toOrdering
+
+      def save(snapshot: GlobalSnapshot): F[Unit] =
+        getLast.flatMap { lastSnapshot =>
+          lastSnapshot.hash match {
+            case Left(error) => error.raiseError[F, Unit]
+            case Right(lastSnapshotHash) =>
+              val expectedLink = (lastSnapshotHash, lastSnapshot.ordinal.next)
+              val actualLink = (snapshot.lastSnapshotHash, snapshot.ordinal)
+              if (expectedLink === actualLink) {
+                ref(snapshot.ordinal).update(_ => snapshot.some).void
+              } else {
+                InvalidGlobalSnapshotChain(expectedLink, actualLink).raiseError[F, Unit]
+              }
+          }
+        }
+
+      def get(ordinal: SnapshotOrdinal): F[Option[GlobalSnapshot]] =
+        ref(ordinal).get
+
+      def getLast: F[GlobalSnapshot] =
+        ref.keys.map(_.sorted).map(_.head).flatMap(k => ref(k).get).map(_.get) // Note: .get is unsafe, cc: @tbekas
```

Comment:
I just copied the interface you created, I'll let you change the methods, depending on your needs

Reply:
The previous implementation was using `NonEmptyList` which guaranteed containing at least one element, hence there was just `GlobalSnapshot` returned instead of `Option[GlobalSnapshot]`. You are relaxing the constraints on the underlying data type to `Map` so you reflect these changes in the interface.

Also as a rule of thumb, I think we should review changes independently of what may or may not be introduced in the future.

Reply:
I can make it a non empty set if you want to use it like that. I just remember you said it will return Option[GlobalSnapshot] anyway